server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoErrorRetryAnalyzer.ts

```diff
@@ -0,0 +1,64 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Lumberjack } from "@fluidframework/server-services-telemetry";
+import { DefaultExceptionRule } from "./defaultExceptionRule";
+import { IMongoExceptionRetryRule } from "./IMongoExceptionRetryRule";
+import { createMongoErrorRetryRuleset } from "./mongoError";
+import { createMongoNetworkErrorRetryRuleset } from "./mongoNetworkError";
+
+export class MongoErrorRetryAnalyzer {
+    private static instance: MongoErrorRetryAnalyzer;
+    private readonly mongoNetworkErrorRetryRuleset: IMongoExceptionRetryRule[];
+    private readonly mongoErrorRetryRuleset: IMongoExceptionRetryRule[];
+    private readonly defaultRule: IMongoExceptionRetryRule;
+
+    public static getInstance(retryRuleOverride: Map<string, boolean>): MongoErrorRetryAnalyzer {
+        if (!this.instance) {
+            this.instance = new MongoErrorRetryAnalyzer(retryRuleOverride);
+        }
+        return this.instance;
+    }
+
+    private constructor(retryRuleOverride: Map<string, boolean>) {
+        this.mongoNetworkErrorRetryRuleset = createMongoNetworkErrorRetryRuleset(retryRuleOverride);
+        this.mongoErrorRetryRuleset = createMongoErrorRetryRuleset(retryRuleOverride);
+        this.defaultRule = new DefaultExceptionRule(retryRuleOverride);
+    }
+
+    shouldRetry(error: Error): boolean {
+        const rule = this.getRetryRule(error);
+        if (!rule) {
+            // This should not happen.
+            Lumberjack.error("MongoErrorRetryAnalyzer.shouldRetry() didn't get a rule", undefined, error);
+            return false;
+        }
+        const ruleName = rule.ruleName;
+        const decision = rule.shouldRetry();
+        const properties = {
+            ruleName,
+            decision,
+        };
+
+        Lumberjack.warning(`Error rule used ${rule.ruleName}, shouldRetry: ${decision}`, properties, error);
```

Comment:
>warning [](http://example.com/codeflow?start=19&length=7)

Why warning?

Reply:
I want to include the error into the loggings, to make sure retry logic working as expected based on exceptions we have. Only `warning` and `error` have the ability to include exception and used `warning`. Also when we have retry used, given the delay it kinds of mean program is in low func mode so used `warning`.